File: haskell/p19.hs
-- P19 Rotate a list N places to the left.

-- Using "splitAt"
f1 :: Int -> [a] -> [a]
f1 _ [] = []
f1 n xs = let (a, b) = splitAt (mod n $ length xs) xs in b ++ a

-- Without using "let"
f2 :: Int -> [a] -> [a]
f2 _ [] = []
f2 n xs = uncurry (flip (++)) $ splitAt (mod n $ length xs) xs

-- Tail recursion
f3 :: Int -> [a] -> [a]
f3 _ [] = []
f3 0 xs = xs
f3 n xs
    | n > 0 = f3 (pred n) (tail xs ++ [head xs])
    | otherwise = f3 (succ n) (last xs : init xs)

File: clojure/P19.clj
; P19 Rotate a list N places to the left.

; Using "split-at"
(defn f1 [n l]
  (if (seq l)
    (let [[a b] (split-at (mod n (count l)) l)]
      (concat b a))
    '()))

; Without using "let"
(defn f2 [n l]
  (if (seq l)
    (apply concat
      (reverse (split-at (mod n (count l)) l)))
    '()))

; Tail recursion
(defn f3 [n l]
  (if (and (seq l) (not (zero? n)))
    (if (pos? n)
      (recur (dec n) (concat (rest l) (list (first l))))
      (recur (inc n) (cons (last l) (drop-last l))))
    l))